File: js/test/fixtures.ts
import { Buffer } from 'node:buffer';
import { createHash } from 'node:crypto';
import type { ModelAttestation } from '../src';
import type { VerifiedTdxQuote } from '../src/types/verification';

export const nonce = '11'.repeat(32);
export const signingAddress = `0x${'22'.repeat(20)}`;
export const tlsFingerprint = '33'.repeat(32);
export const appCompose =
  '{"services":{"model":"example@sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"}}';

type Deferred<T> = {
  promise: Promise<T>;
  resolve: (value: T | PromiseLike<T>) => void;
  reject: (reason?: unknown) => void;
};

export function createDeferred<T>(): Deferred<T> {
  let resolve!: (value: T | PromiseLike<T>) => void;
  let reject!: (reason?: unknown) => void;
  const promise = new Promise<T>((resolvePromise, rejectPromise) => {
    resolve = resolvePromise;
    reject = rejectPromise;
  });
  return { promise, resolve, reject };
}

export function sha256(value: Uint8Array | string): Buffer {
  return createHash('sha256').update(value).digest();
}

export function sha384(value: Uint8Array): Buffer {
  return createHash('sha384').update(value).digest();
}

/** A quote using the Gateway TLS report-data layout. */
export function createGatewayTlsQuote(
  overrides: Partial<VerifiedTdxQuote> = {},
): VerifiedTdxQuote {
  const runtimeDigest = Buffer.alloc(48);
  const rtmr3 = sha384(Buffer.concat([Buffer.alloc(48), runtimeDigest]));
  const reportData = Buffer.concat([
    sha256(
      Buffer.concat([
        Buffer.from(signingAddress.slice(2), 'hex'),
        Buffer.from(tlsFingerprint, 'hex'),
      ]),
    ),
    Buffer.from(nonce, 'hex'),
  ]);
  const mrConfigId = Buffer.concat([
    Buffer.from([0x01]),
    sha256(appCompose),
    Buffer.alloc(15),
  ]);

  return {
    tcbStatus: 'UpToDate',
    advisoryIds: [],
    debugEnabled: false,
    reportData,
    mrConfigId,
    rtMr3: rtmr3,
    ...overrides,
  };
}

/** A Cloud model quote using the signer-and-nonce report-data layout. */
export function createModelQuote(
  overrides: Partial<VerifiedTdxQuote> = {},
): VerifiedTdxQuote {
  return createGatewayTlsQuote({
    reportData: Buffer.concat([
      Buffer.from(signingAddress.slice(2), 'hex'),
      Buffer.alloc(12),
      Buffer.from(nonce, 'hex'),
    ]),
    ...overrides,
  });
}

export function createModelAttestation(
  overrides: Partial<ModelAttestation> = {},
): ModelAttestation {
  return {
    nonce,
    signer: { signingAlgo: 'ecdsa', signingAddress },
    intelQuote: 'aa',
    eventLog: [
      {
        digest: '00'.repeat(48),
        event_type: 0,
        event: 'compose-hash',
        event_payload: 'beef',
        imr: 3,
      },
    ],
    appCompose,
    ...overrides,
  };
}

File: py/test/test_nearai_inference_sdk/fixtures.py
"""Deterministic evidence used by the public SDK contract tests."""

from __future__ import annotations

from dataclasses import replace

from nearai_inference_sdk import (
    GatewayAttestation,
    ModelClientBinding,
    ModelAttestation,
    QuoteVerificationResult,
    SigningIdentity,
)

from nearai_inference_sdk.utils.common import sha256, sha384


NONCE = '11' * 32
MODEL_CLIENT_BINDING = ModelClientBinding(nonce=NONCE)
SIGNING_ADDRESS = f'0x{"22" * 20}'
TLS_FINGERPRINT = '33' * 32
APP_COMPOSE = (
    '{"services":{"model":"example@sha256:'
    'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"}}'
)


def create_gateway_tls_quote(
    *,
    signing_address: str = SIGNING_ADDRESS,
    tls_fingerprint: str = TLS_FINGERPRINT,
    **overrides: object,
) -> QuoteVerificationResult:
    signing_address_bytes = bytes.fromhex(signing_address.removeprefix('0x'))
    report_prefix = sha256(signing_address_bytes + bytes.fromhex(tls_fingerprint))
    quote = QuoteVerificationResult(
        tcb_status='UpToDate',
        advisory_ids=(),
        debug_enabled=False,
        report_data=report_prefix + bytes.fromhex(NONCE),
        mr_config_id=b'\x01' + sha256(APP_COMPOSE.encode()) + bytes(15),
        rt_mr3=sha384(bytes(48) + bytes(48)),
    )
    return replace(quote, **overrides)


def create_model_quote(
    *,
    signing_address: str = SIGNING_ADDRESS,
    **overrides: object,
) -> QuoteVerificationResult:
    signing_address_bytes = bytes.fromhex(signing_address.removeprefix('0x'))
    report_prefix = signing_address_bytes.ljust(32, b'\x00')
    quote = QuoteVerificationResult(
        tcb_status='UpToDate',
        advisory_ids=(),
        debug_enabled=False,
        report_data=report_prefix + bytes.fromhex(NONCE),
        mr_config_id=b'\x01' + sha256(APP_COMPOSE.encode()) + bytes(15),
        rt_mr3=sha384(bytes(48) + bytes(48)),
    )
    return replace(quote, **overrides)


def create_model_attestation(**overrides: object) -> ModelAttestation:
    attestation = ModelAttestation(
        nonce=NONCE,
        signer=SigningIdentity(signing_algo='ecdsa', signing_address=SIGNING_ADDRESS),
        intel_quote='aa',
        event_log=[
            {
                'digest': '00' * 48,
                'event_type': 0,
                'event': 'compose-hash',
                'event_payload': 'beef',
                'imr': 3,
            }
        ],
        app_compose=APP_COMPOSE,
    )
    return replace(attestation, **overrides)


def create_gateway_attestation(**overrides: object) -> GatewayAttestation:
    quote = create_gateway_tls_quote()
    attestation = GatewayAttestation(
        nonce=NONCE,
        signer=SigningIdentity(signing_algo='ecdsa', signing_address=SIGNING_ADDRESS),
        intel_quote='aa',
        event_log=[
            {
                'digest': '00' * 48,
                'event_type': 0,
                'event': 'compose-hash',
                'event_payload': 'beef',
                'imr': 3,
            }
        ],
        app_compose=APP_COMPOSE,
        spki_fingerprint=TLS_FINGERPRINT,
        reported_quote_data=quote.report_data.hex(),
    )
    return replace(attestation, **overrides)
